Sources/SwiftGRPC/Core/ClientNetworkMonitor.swift

```diff
@@ -0,0 +1,162 @@
+            notificationName = .CTServiceRadioAccessTechnologyDidChange
+        } else {
+            notificationName = .CTRadioAccessTechnologyDidChange
+        }
+
+        NotificationCenter.default.addObserver(self, selector: #selector(self.cellularDidChange(_:)),
+                                               name: notificationName, object: nil)
+    }
+
+    @objc
+    private func cellularDidChange(_ notification: NSNotification) {
+        self.queue.async {
+            let newCellularName: String?
+            if #available(iOS 13, *) {
+                let cellularKey = notification.object as? String
+                newCellularName = cellularKey.flatMap { self.cellularInfo.serviceCurrentRadioAccessTechnology?[$0] }
+            } else {
+                newCellularName = notification.object as? String ?? self.cellularInfo.currentRadioAccessTechnology
+            }
+
+            if let newCellularName = newCellularName, self.cellularName != newCellularName {
+                self.cellularName = newCellularName
+                self.callback(State(lastChange: .cellularTechnology(technology: newCellularName),
+                                    isReachable: self.isReachable ?? false))
+            }
+        }
+    }
+
+    // MARK: - Reachability
+
+    private func startMonitoringReachability(host: String) {
+        guard let reachability = SCNetworkReachabilityCreateWithName(nil, host) else {
+            return
+        }
+
+        let info = Unmanaged.passUnretained(self).toOpaque()
+        var context = SCNetworkReachabilityContext(version: 0, info: info, retain: nil,
+                                                   release: nil, copyDescription: nil)
+        let callback: SCNetworkReachabilityCallBack = { _, flags, info in
+            let observer = info.map { Unmanaged<ClientNetworkMonitor>.fromOpaque($0).takeUnretainedValue() }
+            observer?.reachabilityDidChange(with: flags)
+        }
+
+        SCNetworkReachabilitySetCallback(reachability, callback, &context)
+        SCNetworkReachabilityScheduleWithRunLoop(reachability, CFRunLoopGetMain(),
+                                                 CFRunLoopMode.commonModes.rawValue)
+        self.queue.async { [weak self] in
+            self?.reachability = reachability
+
+            var flags = SCNetworkReachabilityFlags()
+            SCNetworkReachabilityGetFlags(reachability, &flags)
+            self?.reachabilityDidChange(with: flags)
+        }
+    }
+
+    private func reachabilityDidChange(with flags: SCNetworkReachabilityFlags) {
+        self.queue.async {
+            let isReachable = flags.contains(.reachable)
+            if let wasReachable = self.isReachable, wasReachable != isReachable {
+                self.callback(State(lastChange: .reachability(isReachable: isReachable), isReachable: isReachable))
+            }
+            self.isReachable = isReachable
+
+            let isUsingWifi = !flags.contains(.isWWAN)
```

Comment:
This might send two callbacks in short succession. Not sure if it matters, just an observation.

Reply:
Yea, I was considering omitting the second one if the first one occurred. Going to update this